docs/orchestration.asciidoc

```diff
@@ -0,0 +1,106 @@
+        resources:
+          requests:
+            storage: 1000Gi
+        storageClassName: standard
+----
+
+The Elasticsearch resource above defines two NodeSets: one for master nodes, using 10Gi volumes, and one for data nodes, using 1000Gi volumes. The Elasticsearch cluster will be composed of 13 nodes: 3 master nodes and 10 data nodes.
+
+[id="{p}-upgrading"]
+==== Upgrading the cluster
+
+ECK handles smooth upgrades from one cluster specification to another. It is possible to apply a new Elasticsearch specification at any time.
+
+Some examples based on the Elasticsearch specification above:
+
+- To add 5 additional data nodes, simply change `count: 10` to `count: 15` in the `data-nodes` NodeSet.
+- To increase the RAM memory limit of data nodes to 32Gi, link:k8s-managing-compute-resources.html[set a different resources limits] in the existing `data-nodes` NodeSet PodTemplate.
+- To replace dedicated master and dedicated data nodes by nodes having both master and data roles, replace the 2 existing NodeSets by a single one with a different name and the corresponding Elasticsearch configuration settings.
+- To upgrade Elasticsearch from version `7.2.0` to `7.3.0`, simply change the value in the `version` field.
+
+No downtime is expected when NodeSet changes are orchestrated by ECK. It ensures that:
+
+- data on a node that is being removed is completely migrated to other nodes before actually removing said node
+- Elasticsearch orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions` are adjusted as necessary for any cluster topology change
+- rolling upgrades are performed safely, reusing PersistentVolumes of the upgraded Elasticsearch nodes
+
+[id="{p}-statefulsets"]
+==== StatefulSets orchestration
+
+Behind the scenes, ECK translates each NodeSet specified in the Elasticsearch resource into a link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[StatefulSet in Kubernetes]. The StatefulSet specification is based on the NodeSet specification:
+
+* `count` corresponds to the number of replicas in the StatefulSet, each replica leading to the creation of a Pod, which corresponds to a single Elasticsearch node
+* `podTemplate` can be used to specify custom settings for the Elasticsearch Pod, overriding the default ones set by ECK on the generated StatefulSet specification
+* the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name. Each Pod will be assigned the StatefulSet name suffixed by an ordinal. The corresponding Elasticsearch node has the same name as the Pod.
+
+The actual Pod creation is handled by the StatefulSet controller in Kubernetes. ECK relies on the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[OnDelete StatefulSet update strategy] since it needs full control over when and how Pods get upgraded to a new revision.
```

Comment:
Is it worth saying that manually deleting a Pod while an upgrade is in progress is a bad idea and should be done to only unlock a situation where the operator does not seem to make any progress ?

Reply:
I think we say this at another place:

> ECK takes the cautionary approach of waiting for green health before progressing but advanced users may force an upgrade by manually deleting pods themselves